how-to/qemu-riscv: fix bugs; fix inconsistence; rm duplicated cloud-init content; refine structure

pzhlkj6612 opened the pull request:
Hello! Thank you all for providing this article. I can run a basic RISC-V machine by reading it.

There are some issues with this article, and I've made some modifications. Please review.

Commit message:

* Overall:
    * All sentences should be ended with a punctuation.
    * All command lines should have a similar structure and arguments to make it easier for learners to discover differences.
    * Other improvements.
* Details:
    * The "cloud-init integration" section has been replaced by the "Cloud-init seed" section since all content in it is covered by the "/how-to/headless-usage" article.
* Tested on:
    * Ubuntu 24.04.
    * Ubuntu 25.04.
    * Ubuntu 25.10-snapshot4.

xypron commented on how-to/qemu-riscv.rst:
```diff
@@ -35,53 +45,56 @@ Using the pre-installed server image
        qemu-system-riscv64 \
            -cpu rva23s64 \
-           -machine virt -nographic -m 2048 -smp 4 \
+           -machine virt -m 4G -smp cpus=2 -nographic \
```
Why do you want 4 GB of memory and only 2 CPUs?

pzhlkj6612 replied:
Well, that's a copy-paste of the last configuration in the docs. So, which one prefer?

- 4 CPUs + 2 GB RAM ([qemu-riscv.rst#L62](https://github.com/canonical/ubuntu-boards-documentation/blame/a76cfbf11b14af5dc21e8ed26e2aa15f8fcd64c9/how-to/qemu-riscv.rst?plain=1#L62))
- 4 CPUs + 4 GB RAM ([qemu-riscv.rst#L110](https://github.com/canonical/ubuntu-boards-documentation/blame/a76cfbf11b14af5dc21e8ed26e2aa15f8fcd64c9/how-to/qemu-riscv.rst?plain=1#L110))
- 2 CPUs + 4 GB RAM ([qemu-riscv.rst#L156](https://github.com/canonical/ubuntu-boards-documentation/blame/a76cfbf11b14af5dc21e8ed26e2aa15f8fcd64c9/how-to/qemu-riscv.rst?plain=1#L156))

xypron commented on how-to/qemu-riscv.rst:
```diff
@@ -35,53 +45,56 @@ Using the pre-installed server image
 
 
-#. Next use u-boot-qemu to boot the virtual machine. A working example with all
+#. Next use U-Boot to boot the virtual machine. A working example with all
```
u-boot-qemu is a package name.

pzhlkj6612 replied:
Hmm, I can't get the point to use a package name here. We have already put the "package - thing" list on the top of this article, so readers should know what "U-Boot" is.

Could you please point me out?

xypron commented on how-to/qemu-riscv.rst:
```diff
@@ -181,19 +194,14 @@ To run your installed Ubuntu image use:
-~~~~~~~~~~~~~~~
-
-Sample files for a cloud-init seed are present on the FAT partition labeled
-"CIDATA". See :doc:`/how-to/headless-usage` for more information.
```
This information is not correct for the installer image and therefore needs to be remove.

pzhlkj6612 commented on how-to/qemu-riscv.rst:
```diff
@@ -22,10 +23,19 @@ The packages can be installed with the following commands:
+   There is a default OpenSBI firmware that is shipped with
+   ``qemu-system-riscv64`` and its dependencies, so you can omit the
+   ``opensbi`` package for simple tests.
+
```
> ... On Debian and Ubuntu OpenSBI is not shipped with QEMU. ...

I must have learned and written something wrong. Those bin files in QEMU deb packages are symbolic links that point to OpenSBI files. My fault. I'll revert it.

> This is only true for upstream QEMU. ...

Actually, I don't know. I also can't find QEMU with OpenSBI-shipped on my own at the moment.

> ... This change was done to allow using a newer OpenSBI.

Yeah, it makes sense.

---

Future people who are interested can take a look at the route I walked.

Symbolic links under "/usr/share/qemu/" that point to OpenSBI files under "/usr/lib/riscv{32,64}-linux-gnu/opensbi/generic/":

- https://git.launchpad.net/ubuntu/+source/qemu/tree/debian/qemu-system-riscv.links?id=b92f7058970ce15bf0e468ff4833f84e3f3732d6
- https://salsa.debian.org/qemu-team/qemu/-/blob/66afb7c4adc42d8df685dc963895b0ef5946937c/debian/qemu-system-riscv.links

The commit about "QEMU stops building OpenSBI":

- https://git.launchpad.net/ubuntu/+source/qemu/commit/debian?id=827cd0d4141f43c2615fce19e316243cb07b7ced
- https://salsa.debian.org/qemu-team/qemu/-/commit/ecd3e6869ac00df121b542dd52698d96fa01d2a5

pzhlkj6612 commented:
Hello @xypron,

Thanks for your detailed review. I've answered questions, fixed what I can fix and rebased onto the newest commit [^1]. Please take a look.

Best regards

Mozi

[^1]: https://github.com/canonical/ubuntu-boards-documentation/compare/f805dca5a79148cf6f159c896b40ac75565f2d89..3469c1f407a8367814c61008d0e0e0ff4ca44669

pzhlkj6612 commented on how-to/qemu-riscv.rst:
```diff
@@ -111,10 +112,11 @@ Running via U-Boot
   .. code-block:: text
 
-      [   68.346028] cloud-init[703]: Cloud-init v. 22.2-0ubuntu1~20.04.3 finished at Thu, 22 Sep 2022 11:35:28 +0000. Datasource DataSourceNoCloud [seed=/var/lib/cloud/seed/nocloud-net][dsmode=net].  Up 68.26 seconds
+      [......] cloud-init[...]: Cloud-init v. ... finished at ...  Up ... seconds
```
@xypron 

> Why don't you want to have a realistic example here? In the proposed form it doesn't look helpful to me.
> That we have to watch out for "Cloud-init finished" is mentioned above.

Thank you! I must not have read the whole paragraph carefully. A realistic example is indeed useful here.

I've performed the undo of my change in 06be42a556339bab5c5b277dbb0590714cd0818d.

Best regards

Mozi
